### PR TITLE
*: fast path point select 

### PR DESCRIPTION
**Pull request by coocood**
## What have you changed? (mandatory)

Skip optimizer and coprocessor for point select.

## What are the type of the changes (mandatory)?

The currently defined types are listed below, please pick one of the types for this PR by removing the others:
- Improvement (non-breaking change which is an improvement to an existing feature)

## How has this PR been tested (mandatory)?

Sysbench point select.

## Benchmark result if necessary (optional)

With a single faketikv and single tidb-server on a 24 core linux machine, improve sysbench point select by 60%, from 19000 to 32000

### CODE REVIEW

**Comment by coocood** [comment]
@tiancaiamao PTAL

File: plan/rule_partition_processor.go [hunk 1]
```diff
@@ -41,7 +41,7 @@ type partitionProcessor struct{}
 func (s *partitionProcessor) optimize(lp LogicalPlan) (LogicalPlan, error) {
 	// NOTE: partitionProcessor will assume all filter conditions are pushed down to
 	// DataSource, there will not be a Selection->DataSource case, so the rewrite just
-	// handle the DataSource node.
+	// Handle the DataSource node.
```
Comment (winoros) [inline plan/rule_partition_processor.go]: It's not the first word of the sentence.

File: plan/stats.go [hunk 2]
```diff
@@ -69,10 +69,14 @@ func newSimpleStats(rowCount float64) *statsInfo {
 	return &statsInfo{count: rowCount}
 }
 
-func (p *basePhysicalPlan) StatsInfo() *statsInfo {
+func (p *basePhysicalPlan) statsInfo() *statsInfo {
```
[the remark this answers is not in the document]
Comment (coocood) [reply plan/stats.go]: The `golint` reports error about returning unexported type.

**Comment by coocood** [comment]
@winoros PTAL

File: plan/point_select_plan.go [hunk 3]
```diff
@@ -0,0 +1,435 @@
...
+			return nil
+		}
+		return append(nvPairs, nameValuePair{colName: colName.Name.Name.L, value: d})
+	}
+	return nil
+}
+
+func findPKHandle(tblInfo *model.TableInfo, pairs []nameValuePair) (d types.Datum) {
+	if !tblInfo.PKIsHandle {
+		return d
```
Comment (tiancaiamao) [inline plan/point_select_plan.go]: d.SetNULL or return something like NULL constant is bettter
This code come with the assumption `types.KindInt64 != 0`

File: types/time.go [hunk 4]
```diff
@@ -2109,26 +2109,26 @@ var monthAbbrev = map[string]gotime.Month{
...
-	"%m": monthNumeric,          // Month, numeric (00..12)
-	"%p": isAMOrPM,              // AM or PM
-	"%r": time12Hour,            // Time, 12-hour (hh:mm:ss followed by AM or PM)
-	"%s": secondsNumeric,        // Seconds (00..59)
-	"%S": secondsNumeric,        // Seconds (00..59)
-	"%T": time24Hour,            // Time, 24-hour (hh:mm:ss)
-	"%Y": yearNumericFourDigits, // Year, numeric, four digits
+	"%b": abbreviatedMonth,           // Abbreviated month name (Jan..Dec)
+	"%c": monthNumeric,               // Month, numeric (0..12)
+	"%d": dayOfMonthNumericTwoDigits, // Day of the month, numeric (00..31)
```
Comment (tiancaiamao) [inline types/time.go]: You can probably do this kind of things in another PR

File: executor/point_select.go [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+
+func (b *executorBuilder) buildPointSelect(p *plan.PointSelectPlan) Executor {
+	return &PointSelectExecutor{
+		ctx:     b.ctx,
+		schema:  p.Schema(),
+		tblInfo: p.TblInfo,
+		idxInfo: p.IndexInfo,
+		idxVals: p.IndexValues,
+		handle:  p.Handle,
+		startTS: b.getStartTS(),
```
[the remark this answers is not in the document]
Comment (coocood) [reply executor/point_select.go]: The statement may be UPDATE/DELETE or in an explicit transaction.

File: executor/point_select.go [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+func (e *PointSelectExecutor) get(key kv.Key) (val []byte, err error) {
+	txn := e.ctx.Txn()
+	if txn != nil && txn.Valid() && !txn.IsReadOnly() {
```
[the remark this answers is not in the document]
Comment (coocood) [reply executor/point_select.go]: Yes

File: types/time.go (hunk 4 above)
Comment (coocood) [reply types/time.go]: This may be caused by merge conflict.

File: plan/point_select_plan.go (hunk 3 above)
Comment (coocood) [reply plan/point_select_plan.go]: uninitialized Datum is null is a very intuitive and reasonable assumption.

**Comment by coocood** [comment]
@tiancaiamao PTAL

File: plan/point_select_plan.go [hunk 7]
```diff
@@ -0,0 +1,435 @@
...
+		Fields:  &ast.FieldList{},
+		From:    updateStmt.TableRefs,
+		Where:   updateStmt.Where,
+		OrderBy: updateStmt.Order,
+		Limit:   updateStmt.Limit,
+	}
+	fastSelect := tryFastSelectPlan(ctx, selStmt)
+	if fastSelect == nil {
+		return nil
+	}
```
Comment (tiancaiamao) [inline plan/point_select_plan.go]: Add `checkFastPlanPrivilege` here ?

File: plan/point_select_plan.go [hunk 8]
```diff
@@ -0,0 +1,435 @@
...
+	for _, checkType := range checkTypes {
+		if !pm.RequestVerification(dbName, fastPlan.TblInfo.Name.L, "", checkType) {
+			return errors.New("privilege check fail")
+		}
+	}
+	return nil
+}
+
+func buildSchemaFromFields(ctx sessionctx.Context, tbl *model.TableInfo, fields []*ast.SelectField) *expression.Schema {
+	dbName := model.NewCIStr(ctx.GetSessionVars().CurrentDB)
```
Comment (lamxTyler) [inline plan/point_select_plan.go]: Why current db?

File: plan/point_select_plan.go [hunk 9]
```diff
@@ -0,0 +1,435 @@
...
+	}
+	fastSelect := tryFastSelectPlan(ctx, selStmt)
+	if fastSelect == nil {
+		return nil
+	}
+	orderedList := buildOrderedList(ctx, fastSelect, updateStmt.List)
+	if orderedList == nil {
+		return nil
+	}
+	udpatePlan := &Update{
```
Comment (lamxTyler) [inline plan/point_select_plan.go]: `udpatePlan` -> `updatePlan`

File: plan/stats.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (winoros) [reply plan/stats.go]: Since it's unexported now, is this interface still needed?

File: plan/point_select_plan.go [hunk 10]
```diff
@@ -0,0 +1,435 @@
...
+	stats       *statsInfo
+}
+
+type nameValuePair struct {
+	colName string
+	value   types.Datum
+}
+
+// Schema implements the Plan interface.
+func (fp *PointSelectPlan) Schema() *expression.Schema {
```
Comment (winoros) [inline plan/point_select_plan.go]: Why named `fp`?

File: plan/point_select_plan.go [hunk 11]
```diff
@@ -0,0 +1,435 @@
...
+		schema := buildSchemaFromFields(ctx, tbl, selStmt.Fields.Fields)
+		if schema == nil {
+			return nil
+		}
+		p := newPointSelectPlan(ctx, schema, tbl)
+		p.Handle = handleDatum.GetInt64()
+		return p
+	}
+	for _, idxInfo := range tbl.Indices {
+		if !idxInfo.Unique {
```
Comment (winoros) [inline plan/point_select_plan.go]: Not unique is also ok?

File: statistics/selectivity_test.go [hunk 12]
```diff
@@ -209,14 +209,11 @@ func (s *testSelectivitySuite) TestPseudoSelectivity(c *C) {
 	testKit.MustExec("drop table if exists t, t1")
 	testKit.MustExec("create table t(a int, b int, unique key idx(a,b))")
 	testKit.MustQuery("explain select * from t where a = 1 and b = 1").Check(testkit.Rows(
-		"IndexReader_6 1.00 root index:IndexScan_5",
```
Comment (winoros) [inline statistics/selectivity_test.go]: Since it's no longer call `Selectivity`, this case could be just removed.

File: statistics/selectivity_test.go [hunk 13]
```diff
@@ -209,14 +209,11 @@ func (s *testSelectivitySuite) TestPseudoSelectivity(c *C) {
...
-		"└─IndexScan_5 1.00 cop table:t, index:a, b, range:[1 1,1 1], keep order:false, stats:pseudo"))
+		"Point_Select_1 1.00 root Point_Select"))
 
 	testKit.MustExec("create table t1(a int, b int, primary key(a))")
 	testKit.MustQuery("explain select b from t1 where a = 1").Check(testkit.Rows(
-		"Projection_4 1.00 root test.t1.b",
-		"└─TableReader_6 1.00 root data:TableScan_5",
-		"  └─TableScan_5 1.00 cop table:t1, range:[1,1], keep order:false, stats:pseudo"))
+		"Point_Select_1 1.00 root Point_Select"))
```
Comment (winoros) [inline statistics/selectivity_test.go]: ditto.

File: plan/point_select_plan.go [hunk 14]
```diff
@@ -0,0 +1,435 @@
...
+}
+
+// ToPB converts physical plan to tipb executor.
+func (fp *PointSelectPlan) ToPB(ctx sessionctx.Context) (*tipb.Executor, error) {
+	return nil, nil
+}
+
+// ExplainInfo returns operator information to be explained.
+func (fp *PointSelectPlan) ExplainInfo() string {
+	return "Point_Select"
```
Comment (zz-jason) [inline plan/point_select_plan.go]: we should give more info instead of just `Point_Select`

File: plan/plan.go [hunk 15]
```diff
@@ -229,6 +229,9 @@ type PhysicalPlan interface {
 	// getChildReqProps gets the required property by child index.
 	getChildReqProps(idx int) *requiredProp
 
+	// StatsCount returns the count of statsInfo for this plan.
+	StatsCount() float64
```
Comment (zz-jason) [inline plan/plan.go]: I think this function can be removed.

File: plan/point_select_plan.go [hunk 16]
```diff
@@ -0,0 +1,435 @@
...
+	"github.com/pingcap/tidb/model"
+	"github.com/pingcap/tidb/mysql"
+	"github.com/pingcap/tidb/parser/opcode"
+	"github.com/pingcap/tidb/privilege"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tipb/go-tipb"
+)
+
+// PointSelectPlan is a fast plan for point select.
```
Comment (zz-jason) [inline plan/point_select_plan.go]: we need to add more comments about this type of plan node:
1. what is this operator used for?
2. when will this operator being used?

File: plan/point_select_plan.go [hunk 17]
```diff
@@ -0,0 +1,435 @@
...
+	return nil
+}
+
+// SetChildren sets the children for the plan.
+func (fp *PointSelectPlan) SetChildren(...PhysicalPlan) {}
+
+// ResolveIndices resolves the indices for columns. After doing this, the columns can evaluate the rows by their indices.
+func (fp *PointSelectPlan) ResolveIndices() {}
+
+func tryFastPlan(ctx sessionctx.Context, node ast.Node) Plan {
```
Comment (zz-jason) [inline plan/point_select_plan.go]: add more comment about this function.

File: plan/point_select_plan.go [hunk 18]
```diff
@@ -0,0 +1,435 @@
...
+		p.IndexInfo = idxInfo
+		p.IndexValues = idxValues
+		return p
+	}
+	return nil
+}
+
+func newPointSelectPlan(ctx sessionctx.Context, schema *expression.Schema, tbl *model.TableInfo) *PointSelectPlan {
+	p := &PointSelectPlan{
+		basePlan: newBasePlan(ctx, "Point_Select"),
```
Comment (zz-jason) [inline plan/point_select_plan.go]: I think "PointGet" is a better name, and we can put this type together with other operator types.

File: plan/point_select_plan.go [hunk 19]
```diff
@@ -0,0 +1,435 @@
...
+		return nil
+	}
+	tblName, ok := tblSrc.Source.(*ast.TableName)
+	if !ok {
+		return nil
+	}
+	return tblName
+}
+
+func getNameValuePairs(nvPairs []nameValuePair, expr ast.ExprNode) []nameValuePair {
```
Comment (zz-jason) [inline plan/point_select_plan.go]: add more comments for this function, for example: what kind of expr functions this functions handles?

File: session/session.go [hunk 20]
```diff
@@ -277,6 +277,9 @@ func (s *session) doCommit(ctx context.Context) error {
 		s.txn.changeToInvalid()
 		s.sessionVars.SetStatusFlag(mysql.ServerStatusInTrans, false)
 	}()
+	if s.txn.IsReadOnly() {
```
Comment (jackysp) [inline session/session.go]: Is this a bug fix?

File: store/tikv/txn.go [hunk 21]
```diff
@@ -245,6 +245,7 @@ func (txn *tikvTxn) LockKeys(keys ...kv.Key) error {
 	for _, key := range keys {
 		txn.lockKeys = append(txn.lockKeys, key)
 	}
+	txn.dirty = true
```
Comment (jackysp) [inline store/tikv/txn.go]: Is this a bug fix?

File: plan/point_select_plan.go [hunk 22]
```diff
@@ -0,0 +1,435 @@
...
+
+// SetChildren sets the children for the plan.
+func (fp *PointSelectPlan) SetChildren(...PhysicalPlan) {}
+
+// ResolveIndices resolves the indices for columns. After doing this, the columns can evaluate the rows by their indices.
+func (fp *PointSelectPlan) ResolveIndices() {}
+
+func tryFastPlan(ctx sessionctx.Context, node ast.Node) Plan {
+	if PreparedPlanCacheEnabled {
+		// plan cache not supported yet.
```
Comment (jackysp) [inline plan/point_select_plan.go]: Is there any plan to support fast-path with plan cache?

**Comment by jackysp** [comment]
I think it's better to find a consistent name for it, maybe PointXXX or FastPathXXX?
:-)

File: plan/point_select_plan.go (hunk 22 above)
Comment (coocood) [reply plan/point_select_plan.go]: No

File: session/session.go (hunk 20 above)
Comment (coocood) [reply session/session.go]: It's an optimization.

File: store/tikv/txn.go (hunk 21 above)
Comment (coocood) [reply store/tikv/txn.go]: If we add the optimization that return early if read only, then this is needed.

File: plan/point_select_plan.go (hunk 8 above)
Comment (coocood) [reply plan/point_select_plan.go]: We should pass in the DB name.

File: plan/point_select_plan.go (hunk 11 above)
Comment (coocood) [reply plan/point_select_plan.go]: We only care about the unique index.

File: plan/point_select_plan.go (hunk 8 above)
Comment (lamxTyler) [reply plan/point_select_plan.go]: But it is not necessarily the current db, for example, `test.t`.

File: store/tikv/txn.go (hunk 21 above)
Comment (jackysp) [reply store/tikv/txn.go]: got it

File: plan/point_select_plan.go (hunk 22 above)
Comment (jackysp) [reply plan/point_select_plan.go]: I think the comment could be improved.

File: executor/point_select.go [hunk 23]
```diff
@@ -0,0 +1,186 @@
...
+		encodedIdxVals, err := codec.EncodeKey(e.ctx.GetSessionVars().StmtCtx, nil, e.idxVals...)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		idxKey := tablecodec.EncodeIndexSeekKey(e.tblInfo.ID, e.idxInfo.ID, encodedIdxVals)
+		handleVal, err := e.get(idxKey)
+		if err != nil && !kv.ErrNotExist.Equal(err) {
+			return errors.Trace(err)
+		}
+		if len(handleVal) == 0 {
```
[the remark this answers is not in the document]
Comment (coocood) [reply executor/point_select.go]: When the key is deleted, it can be found in mem buffer but `handleVal` length is 0.

File: plan/plan.go (hunk 15 above)
Comment (coocood) [reply plan/plan.go]: This is called by `executor` package to determine if the query is expensive.

**Comment by ngaut** [comment]
Yes, we should keep consistent.

File: executor/point_get.go [hunk 24]
```diff
@@ -0,0 +1,197 @@
...
+	"github.com/pingcap/tidb/table"
+	"github.com/pingcap/tidb/table/tables"
+	"github.com/pingcap/tidb/tablecodec"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/codec"
+	"golang.org/x/net/context"
+)
+
+func (b *executorBuilder) buildPointSelect(p *plan.PointGetPlan) Executor {
```
Comment (ngaut) [inline executor/point_get.go]: s/buildPointSelect/buildPointGet/

File: executor/point_get.go [hunk 25]
```diff
@@ -0,0 +1,197 @@
...
+func (e *PointGetExecutor) get(key kv.Key) (val []byte, err error) {
+	txn := e.ctx.Txn()
+	if txn != nil && txn.Valid() && !txn.IsReadOnly() {
+		return txn.Get(key)
+	}
+	return e.snapshot.Get(key)
+}
+
+func (e *PointGetExecutor) decodeRowValToChunk(rowVal []byte, chk *chunk.Chunk) error {
+	colIDs := make(map[int64]int)
```
Comment (ngaut) [inline executor/point_get.go]: Make with capacity.

**Comment by coocood** [comment]
@tiancaiamao @winoros @zz-jason @jackysp 

I remove the point-get plan for `UPDATE` for now, to make the test pass quickly.

Update support can be added in future PR.

**Comment by tiancaiamao** [comment]
LGTM

File: plan/point_get_plan.go [hunk 26]
```diff
@@ -0,0 +1,423 @@
...
+			return nil
+		}
+	}
+	pairs := make([]nameValuePair, 0, 4)
+	pairs = getNameValuePairs(pairs, selStmt.Where)
+	if pairs == nil {
+		return nil
+	}
+	handleDatum := findPKHandle(tbl, pairs)
+	if handleDatum.Kind() == types.KindInt64 {
```
Comment (shenli) [inline plan/point_get_plan.go]: Consider unsign flag?

File: executor/builder.go [hunk 27]
```diff
@@ -1483,7 +1485,7 @@ func buildNoRangeTableReader(b *executorBuilder, v *plan.PhysicalTableReader) (*
 	if containsLimit(dagReq.Executors) {
 		e.feedback = statistics.NewQueryFeedback(0, nil, 0, ts.Desc)
 	} else {
-		e.feedback = statistics.NewQueryFeedback(ts.Table.ID, ts.Hist, ts.StatsInfo().Count(), ts.Desc)
+		e.feedback = statistics.NewQueryFeedback(ts.Table.ID, ts.Hist, int64(ts.StatsCount()), ts.Desc)
```
Comment (zz-jason) [inline executor/builder.go]: why make this change?

Comment (coocood) [reply executor/builder.go]: go lint reports error because it returns an unexported type `statsInfo`.

File: plan/point_get_plan.go (hunk 26 above)
Comment (coocood) [reply plan/point_get_plan.go]: `handle` is always int64.
unsign flag doesn't matter.
